// @ts-safeql/eslint-plugin@2.0.3/lib/utils/ts.utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TSUtils = void 0;
const typescript_1 = __importDefault(require("typescript"));
exports.TSUtils = {
    isTypeUnion: (typeNode) => {
        return (typeNode === null || typeNode === void 0 ? void 0 : typeNode.kind) === typescript_1.default.SyntaxKind.UnionType;
    },
    isTsUnionType(type) {
        return type.flags === typescript_1.default.TypeFlags.Union;
    },
    isTsTypeReference(type) {
        return exports.TSUtils.isTsObjectType(type) && type.objectFlags === typescript_1.default.ObjectFlags.Reference;
    },
    isTsArrayUnionType(checker, type) {
        if (!exports.TSUtils.isTsTypeReference(type)) {
            return false;
        }
        const firstArgument = checker.getTypeArguments(type)[0];
        if (firstArgument === undefined) {
            return false;
        }
        return exports.TSUtils.isTsUnionType(firstArgument);
    },
    isTsObjectType(type) {
        return type.flags === typescript_1.default.TypeFlags.Object;
    },
};
