// @ts-safeql/shared@0.2.0/lib/parsedQuery.js
"use strict";
/* eslint-disable @typescript-eslint/no-empty-interface  */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-namespace */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LibPgQueryAST = void 0;
// https://raw.githubusercontent.com/pganalyze/libpg_query/f1b475ddcbea7951a7ef825f137abdc54e5a34e4/protobuf/pg_query.proto
var LibPgQueryAST;
(function (LibPgQueryAST) {
    let OverridingKind;
    (function (OverridingKind) {
        OverridingKind[OverridingKind["OVERRIDING_KIND_UNDEFINED"] = 0] = "OVERRIDING_KIND_UNDEFINED";
        OverridingKind[OverridingKind["OVERRIDING_NOT_SET"] = 1] = "OVERRIDING_NOT_SET";
        OverridingKind[OverridingKind["OVERRIDING_USER_VALUE"] = 2] = "OVERRIDING_USER_VALUE";
        OverridingKind[OverridingKind["OVERRIDING_SYSTEM_VALUE"] = 3] = "OVERRIDING_SYSTEM_VALUE";
        OverridingKind[OverridingKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(OverridingKind || (OverridingKind = {}));
    let QuerySource;
    (function (QuerySource) {
        QuerySource[QuerySource["QUERY_SOURCE_UNDEFINED"] = 0] = "QUERY_SOURCE_UNDEFINED";
        QuerySource[QuerySource["QSRC_ORIGINAL"] = 1] = "QSRC_ORIGINAL";
        QuerySource[QuerySource["QSRC_PARSER"] = 2] = "QSRC_PARSER";
        QuerySource[QuerySource["QSRC_INSTEAD_RULE"] = 3] = "QSRC_INSTEAD_RULE";
        QuerySource[QuerySource["QSRC_QUAL_INSTEAD_RULE"] = 4] = "QSRC_QUAL_INSTEAD_RULE";
        QuerySource[QuerySource["QSRC_NON_INSTEAD_RULE"] = 5] = "QSRC_NON_INSTEAD_RULE";
        QuerySource[QuerySource["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(QuerySource || (QuerySource = {}));
    let SortByDir;
    (function (SortByDir) {
        SortByDir[SortByDir["SORT_BY_DIR_UNDEFINED"] = 0] = "SORT_BY_DIR_UNDEFINED";
        SortByDir[SortByDir["SORTBY_DEFAULT"] = 1] = "SORTBY_DEFAULT";
        SortByDir[SortByDir["SORTBY_ASC"] = 2] = "SORTBY_ASC";
        SortByDir[SortByDir["SORTBY_DESC"] = 3] = "SORTBY_DESC";
        SortByDir[SortByDir["SORTBY_USING"] = 4] = "SORTBY_USING";
        SortByDir[SortByDir["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(SortByDir || (SortByDir = {}));
    let SortByNulls;
    (function (SortByNulls) {
        SortByNulls[SortByNulls["SORT_BY_NULLS_UNDEFINED"] = 0] = "SORT_BY_NULLS_UNDEFINED";
        SortByNulls[SortByNulls["SORTBY_NULLS_DEFAULT"] = 1] = "SORTBY_NULLS_DEFAULT";
        SortByNulls[SortByNulls["SORTBY_NULLS_FIRST"] = 2] = "SORTBY_NULLS_FIRST";
        SortByNulls[SortByNulls["SORTBY_NULLS_LAST"] = 3] = "SORTBY_NULLS_LAST";
        SortByNulls[SortByNulls["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(SortByNulls || (SortByNulls = {}));
    let AExprKind;
    (function (AExprKind) {
        AExprKind["A_EXPR_KIND_UNDEFINED"] = "A_EXPR_KIND_UNDEFINED";
        AExprKind["AEXPR_OP"] = "AEXPR_OP";
        AExprKind["AEXPR_OP_ANY"] = "AEXPR_OP_ANY";
        AExprKind["AEXPR_OP_ALL"] = "AEXPR_OP_ALL";
        AExprKind["AEXPR_DISTINCT"] = "AEXPR_DISTINCT";
        AExprKind["AEXPR_NOT_DISTINCT"] = "AEXPR_NOT_DISTINCT";
        AExprKind["AEXPR_NULLIF"] = "AEXPR_NULLIF";
        AExprKind["AEXPR_IN"] = "AEXPR_IN";
        AExprKind["AEXPR_LIKE"] = "AEXPR_LIKE";
        AExprKind["AEXPR_ILIKE"] = "AEXPR_ILIKE";
        AExprKind["AEXPR_SIMILAR"] = "AEXPR_SIMILAR";
        AExprKind["AEXPR_BETWEEN"] = "AEXPR_BETWEEN";
        AExprKind["AEXPR_NOT_BETWEEN"] = "AEXPR_NOT_BETWEEN";
        AExprKind["AEXPR_BETWEEN_SYM"] = "AEXPR_BETWEEN_SYM";
        AExprKind["AEXPR_NOT_BETWEEN_SYM"] = "AEXPR_NOT_BETWEEN_SYM";
        AExprKind["UNRECOGNIZED"] = "UNRECOGNIZED";
    })(AExprKind = LibPgQueryAST.AExprKind || (LibPgQueryAST.AExprKind = {}));
    let RoleSpecType;
    (function (RoleSpecType) {
        RoleSpecType[RoleSpecType["ROLE_SPEC_TYPE_UNDEFINED"] = 0] = "ROLE_SPEC_TYPE_UNDEFINED";
        RoleSpecType[RoleSpecType["ROLESPEC_CSTRING"] = 1] = "ROLESPEC_CSTRING";
        RoleSpecType[RoleSpecType["ROLESPEC_CURRENT_ROLE"] = 2] = "ROLESPEC_CURRENT_ROLE";
        RoleSpecType[RoleSpecType["ROLESPEC_CURRENT_USER"] = 3] = "ROLESPEC_CURRENT_USER";
        RoleSpecType[RoleSpecType["ROLESPEC_SESSION_USER"] = 4] = "ROLESPEC_SESSION_USER";
        RoleSpecType[RoleSpecType["ROLESPEC_PUBLIC"] = 5] = "ROLESPEC_PUBLIC";
        RoleSpecType[RoleSpecType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(RoleSpecType || (RoleSpecType = {}));
    let DefElemAction;
    (function (DefElemAction) {
        DefElemAction[DefElemAction["DEF_ELEM_ACTION_UNDEFINED"] = 0] = "DEF_ELEM_ACTION_UNDEFINED";
        DefElemAction[DefElemAction["DEFELEM_UNSPEC"] = 1] = "DEFELEM_UNSPEC";
        DefElemAction[DefElemAction["DEFELEM_SET"] = 2] = "DEFELEM_SET";
        DefElemAction[DefElemAction["DEFELEM_ADD"] = 3] = "DEFELEM_ADD";
        DefElemAction[DefElemAction["DEFELEM_DROP"] = 4] = "DEFELEM_DROP";
        DefElemAction[DefElemAction["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(DefElemAction || (DefElemAction = {}));
    let PartitionRangeDatumKind;
    (function (PartitionRangeDatumKind) {
        PartitionRangeDatumKind[PartitionRangeDatumKind["PARTITION_RANGE_DATUM_KIND_UNDEFINED"] = 0] = "PARTITION_RANGE_DATUM_KIND_UNDEFINED";
        PartitionRangeDatumKind[PartitionRangeDatumKind["PARTITION_RANGE_DATUM_MINVALUE"] = 1] = "PARTITION_RANGE_DATUM_MINVALUE";
        PartitionRangeDatumKind[PartitionRangeDatumKind["PARTITION_RANGE_DATUM_VALUE"] = 2] = "PARTITION_RANGE_DATUM_VALUE";
        PartitionRangeDatumKind[PartitionRangeDatumKind["PARTITION_RANGE_DATUM_MAXVALUE"] = 3] = "PARTITION_RANGE_DATUM_MAXVALUE";
        PartitionRangeDatumKind[PartitionRangeDatumKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(PartitionRangeDatumKind || (PartitionRangeDatumKind = {}));
    let RTEKind;
    (function (RTEKind) {
        RTEKind[RTEKind["RTEKIND_UNDEFINED"] = 0] = "RTEKIND_UNDEFINED";
        RTEKind[RTEKind["RTE_RELATION"] = 1] = "RTE_RELATION";
        RTEKind[RTEKind["RTE_SUBQUERY"] = 2] = "RTE_SUBQUERY";
        RTEKind[RTEKind["RTE_JOIN"] = 3] = "RTE_JOIN";
        RTEKind[RTEKind["RTE_FUNCTION"] = 4] = "RTE_FUNCTION";
        RTEKind[RTEKind["RTE_TABLEFUNC"] = 5] = "RTE_TABLEFUNC";
        RTEKind[RTEKind["RTE_VALUES"] = 6] = "RTE_VALUES";
        RTEKind[RTEKind["RTE_CTE"] = 7] = "RTE_CTE";
        RTEKind[RTEKind["RTE_NAMEDTUPLESTORE"] = 8] = "RTE_NAMEDTUPLESTORE";
        RTEKind[RTEKind["RTE_RESULT"] = 9] = "RTE_RESULT";
        RTEKind[RTEKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(RTEKind || (RTEKind = {}));
    let WCOKind;
    (function (WCOKind) {
        WCOKind[WCOKind["WCOKIND_UNDEFINED"] = 0] = "WCOKIND_UNDEFINED";
        WCOKind[WCOKind["WCO_VIEW_CHECK"] = 1] = "WCO_VIEW_CHECK";
        WCOKind[WCOKind["WCO_RLS_INSERT_CHECK"] = 2] = "WCO_RLS_INSERT_CHECK";
        WCOKind[WCOKind["WCO_RLS_UPDATE_CHECK"] = 3] = "WCO_RLS_UPDATE_CHECK";
        WCOKind[WCOKind["WCO_RLS_CONFLICT_CHECK"] = 4] = "WCO_RLS_CONFLICT_CHECK";
        WCOKind[WCOKind["WCO_RLS_MERGE_UPDATE_CHECK"] = 5] = "WCO_RLS_MERGE_UPDATE_CHECK";
        WCOKind[WCOKind["WCO_RLS_MERGE_DELETE_CHECK"] = 6] = "WCO_RLS_MERGE_DELETE_CHECK";
        WCOKind[WCOKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(WCOKind || (WCOKind = {}));
    let GroupingSetKind;
    (function (GroupingSetKind) {
        GroupingSetKind[GroupingSetKind["GROUPING_SET_KIND_UNDEFINED"] = 0] = "GROUPING_SET_KIND_UNDEFINED";
        GroupingSetKind[GroupingSetKind["GROUPING_SET_EMPTY"] = 1] = "GROUPING_SET_EMPTY";
        GroupingSetKind[GroupingSetKind["GROUPING_SET_SIMPLE"] = 2] = "GROUPING_SET_SIMPLE";
        GroupingSetKind[GroupingSetKind["GROUPING_SET_ROLLUP"] = 3] = "GROUPING_SET_ROLLUP";
        GroupingSetKind[GroupingSetKind["GROUPING_SET_CUBE"] = 4] = "GROUPING_SET_CUBE";
        GroupingSetKind[GroupingSetKind["GROUPING_SET_SETS"] = 5] = "GROUPING_SET_SETS";
        GroupingSetKind[GroupingSetKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(GroupingSetKind || (GroupingSetKind = {}));
    let CTEMaterialize;
    (function (CTEMaterialize) {
        CTEMaterialize[CTEMaterialize["CTEMATERIALIZE_UNDEFINED"] = 0] = "CTEMATERIALIZE_UNDEFINED";
        CTEMaterialize[CTEMaterialize["CTEMaterializeDefault"] = 1] = "CTEMaterializeDefault";
        CTEMaterialize[CTEMaterialize["CTEMaterializeAlways"] = 2] = "CTEMaterializeAlways";
        CTEMaterialize[CTEMaterialize["CTEMaterializeNever"] = 3] = "CTEMaterializeNever";
        CTEMaterialize[CTEMaterialize["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(CTEMaterialize || (CTEMaterialize = {}));
    let SetOperation;
    (function (SetOperation) {
        SetOperation[SetOperation["SET_OPERATION_UNDEFINED"] = 0] = "SET_OPERATION_UNDEFINED";
        SetOperation[SetOperation["SETOP_NONE"] = 1] = "SETOP_NONE";
        SetOperation[SetOperation["SETOP_UNION"] = 2] = "SETOP_UNION";
        SetOperation[SetOperation["SETOP_INTERSECT"] = 3] = "SETOP_INTERSECT";
        SetOperation[SetOperation["SETOP_EXCEPT"] = 4] = "SETOP_EXCEPT";
        SetOperation[SetOperation["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(SetOperation || (SetOperation = {}));
    let ObjectType;
    (function (ObjectType) {
        ObjectType[ObjectType["OBJECT_TYPE_UNDEFINED"] = 0] = "OBJECT_TYPE_UNDEFINED";
        ObjectType[ObjectType["OBJECT_ACCESS_METHOD"] = 1] = "OBJECT_ACCESS_METHOD";
        ObjectType[ObjectType["OBJECT_AGGREGATE"] = 2] = "OBJECT_AGGREGATE";
        ObjectType[ObjectType["OBJECT_AMOP"] = 3] = "OBJECT_AMOP";
        ObjectType[ObjectType["OBJECT_AMPROC"] = 4] = "OBJECT_AMPROC";
        ObjectType[ObjectType["OBJECT_ATTRIBUTE"] = 5] = "OBJECT_ATTRIBUTE";
        ObjectType[ObjectType["OBJECT_CAST"] = 6] = "OBJECT_CAST";
        ObjectType[ObjectType["OBJECT_COLUMN"] = 7] = "OBJECT_COLUMN";
        ObjectType[ObjectType["OBJECT_COLLATION"] = 8] = "OBJECT_COLLATION";
        ObjectType[ObjectType["OBJECT_CONVERSION"] = 9] = "OBJECT_CONVERSION";
        ObjectType[ObjectType["OBJECT_DATABASE"] = 10] = "OBJECT_DATABASE";
        ObjectType[ObjectType["OBJECT_DEFAULT"] = 11] = "OBJECT_DEFAULT";
        ObjectType[ObjectType["OBJECT_DEFACL"] = 12] = "OBJECT_DEFACL";
        ObjectType[ObjectType["OBJECT_DOMAIN"] = 13] = "OBJECT_DOMAIN";
        ObjectType[ObjectType["OBJECT_DOMCONSTRAINT"] = 14] = "OBJECT_DOMCONSTRAINT";
        ObjectType[ObjectType["OBJECT_EVENT_TRIGGER"] = 15] = "OBJECT_EVENT_TRIGGER";
        ObjectType[ObjectType["OBJECT_EXTENSION"] = 16] = "OBJECT_EXTENSION";
        ObjectType[ObjectType["OBJECT_FDW"] = 17] = "OBJECT_FDW";
        ObjectType[ObjectType["OBJECT_FOREIGN_SERVER"] = 18] = "OBJECT_FOREIGN_SERVER";
        ObjectType[ObjectType["OBJECT_FOREIGN_TABLE"] = 19] = "OBJECT_FOREIGN_TABLE";
        ObjectType[ObjectType["OBJECT_FUNCTION"] = 20] = "OBJECT_FUNCTION";
        ObjectType[ObjectType["OBJECT_INDEX"] = 21] = "OBJECT_INDEX";
        ObjectType[ObjectType["OBJECT_LANGUAGE"] = 22] = "OBJECT_LANGUAGE";
        ObjectType[ObjectType["OBJECT_LARGEOBJECT"] = 23] = "OBJECT_LARGEOBJECT";
        ObjectType[ObjectType["OBJECT_MATVIEW"] = 24] = "OBJECT_MATVIEW";
        ObjectType[ObjectType["OBJECT_OPCLASS"] = 25] = "OBJECT_OPCLASS";
        ObjectType[ObjectType["OBJECT_OPERATOR"] = 26] = "OBJECT_OPERATOR";
        ObjectType[ObjectType["OBJECT_OPFAMILY"] = 27] = "OBJECT_OPFAMILY";
        ObjectType[ObjectType["OBJECT_PARAMETER_ACL"] = 28] = "OBJECT_PARAMETER_ACL";
        ObjectType[ObjectType["OBJECT_POLICY"] = 29] = "OBJECT_POLICY";
        ObjectType[ObjectType["OBJECT_PROCEDURE"] = 30] = "OBJECT_PROCEDURE";
        ObjectType[ObjectType["OBJECT_PUBLICATION"] = 31] = "OBJECT_PUBLICATION";
        ObjectType[ObjectType["OBJECT_PUBLICATION_NAMESPACE"] = 32] = "OBJECT_PUBLICATION_NAMESPACE";
        ObjectType[ObjectType["OBJECT_PUBLICATION_REL"] = 33] = "OBJECT_PUBLICATION_REL";
        ObjectType[ObjectType["OBJECT_ROLE"] = 34] = "OBJECT_ROLE";
        ObjectType[ObjectType["OBJECT_ROUTINE"] = 35] = "OBJECT_ROUTINE";
        ObjectType[ObjectType["OBJECT_RULE"] = 36] = "OBJECT_RULE";
        ObjectType[ObjectType["OBJECT_SCHEMA"] = 37] = "OBJECT_SCHEMA";
        ObjectType[ObjectType["OBJECT_SEQUENCE"] = 38] = "OBJECT_SEQUENCE";
        ObjectType[ObjectType["OBJECT_SUBSCRIPTION"] = 39] = "OBJECT_SUBSCRIPTION";
        ObjectType[ObjectType["OBJECT_STATISTIC_EXT"] = 40] = "OBJECT_STATISTIC_EXT";
        ObjectType[ObjectType["OBJECT_TABCONSTRAINT"] = 41] = "OBJECT_TABCONSTRAINT";
        ObjectType[ObjectType["OBJECT_TABLE"] = 42] = "OBJECT_TABLE";
        ObjectType[ObjectType["OBJECT_TABLESPACE"] = 43] = "OBJECT_TABLESPACE";
        ObjectType[ObjectType["OBJECT_TRANSFORM"] = 44] = "OBJECT_TRANSFORM";
        ObjectType[ObjectType["OBJECT_TRIGGER"] = 45] = "OBJECT_TRIGGER";
        ObjectType[ObjectType["OBJECT_TSCONFIGURATION"] = 46] = "OBJECT_TSCONFIGURATION";
        ObjectType[ObjectType["OBJECT_TSDICTIONARY"] = 47] = "OBJECT_TSDICTIONARY";
        ObjectType[ObjectType["OBJECT_TSPARSER"] = 48] = "OBJECT_TSPARSER";
        ObjectType[ObjectType["OBJECT_TSTEMPLATE"] = 49] = "OBJECT_TSTEMPLATE";
        ObjectType[ObjectType["OBJECT_TYPE"] = 50] = "OBJECT_TYPE";
        ObjectType[ObjectType["OBJECT_USER_MAPPING"] = 51] = "OBJECT_USER_MAPPING";
        ObjectType[ObjectType["OBJECT_VIEW"] = 52] = "OBJECT_VIEW";
        ObjectType[ObjectType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(ObjectType || (ObjectType = {}));
    let DropBehavior;
    (function (DropBehavior) {
        DropBehavior[DropBehavior["DROP_BEHAVIOR_UNDEFINED"] = 0] = "DROP_BEHAVIOR_UNDEFINED";
        DropBehavior[DropBehavior["DROP_RESTRICT"] = 1] = "DROP_RESTRICT";
        DropBehavior[DropBehavior["DROP_CASCADE"] = 2] = "DROP_CASCADE";
        DropBehavior[DropBehavior["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(DropBehavior || (DropBehavior = {}));
    let AlterTableType;
    (function (AlterTableType) {
        AlterTableType[AlterTableType["ALTER_TABLE_TYPE_UNDEFINED"] = 0] = "ALTER_TABLE_TYPE_UNDEFINED";
        AlterTableType[AlterTableType["AT_AddColumn"] = 1] = "AT_AddColumn";
        AlterTableType[AlterTableType["AT_AddColumnRecurse"] = 2] = "AT_AddColumnRecurse";
        AlterTableType[AlterTableType["AT_AddColumnToView"] = 3] = "AT_AddColumnToView";
        AlterTableType[AlterTableType["AT_ColumnDefault"] = 4] = "AT_ColumnDefault";
        AlterTableType[AlterTableType["AT_CookedColumnDefault"] = 5] = "AT_CookedColumnDefault";
        AlterTableType[AlterTableType["AT_DropNotNull"] = 6] = "AT_DropNotNull";
        AlterTableType[AlterTableType["AT_SetNotNull"] = 7] = "AT_SetNotNull";
        AlterTableType[AlterTableType["AT_DropExpression"] = 8] = "AT_DropExpression";
        AlterTableType[AlterTableType["AT_CheckNotNull"] = 9] = "AT_CheckNotNull";
        AlterTableType[AlterTableType["AT_SetStatistics"] = 10] = "AT_SetStatistics";
        AlterTableType[AlterTableType["AT_SetOptions"] = 11] = "AT_SetOptions";
        AlterTableType[AlterTableType["AT_ResetOptions"] = 12] = "AT_ResetOptions";
        AlterTableType[AlterTableType["AT_SetStorage"] = 13] = "AT_SetStorage";
        AlterTableType[AlterTableType["AT_SetCompression"] = 14] = "AT_SetCompression";
        AlterTableType[AlterTableType["AT_DropColumn"] = 15] = "AT_DropColumn";
        AlterTableType[AlterTableType["AT_DropColumnRecurse"] = 16] = "AT_DropColumnRecurse";
        AlterTableType[AlterTableType["AT_AddIndex"] = 17] = "AT_AddIndex";
        AlterTableType[AlterTableType["AT_ReAddIndex"] = 18] = "AT_ReAddIndex";
        AlterTableType[AlterTableType["AT_AddConstraint"] = 19] = "AT_AddConstraint";
        AlterTableType[AlterTableType["AT_AddConstraintRecurse"] = 20] = "AT_AddConstraintRecurse";
        AlterTableType[AlterTableType["AT_ReAddConstraint"] = 21] = "AT_ReAddConstraint";
        AlterTableType[AlterTableType["AT_ReAddDomainConstraint"] = 22] = "AT_ReAddDomainConstraint";
        AlterTableType[AlterTableType["AT_AlterConstraint"] = 23] = "AT_AlterConstraint";
        AlterTableType[AlterTableType["AT_ValidateConstraint"] = 24] = "AT_ValidateConstraint";
        AlterTableType[AlterTableType["AT_ValidateConstraintRecurse"] = 25] = "AT_ValidateConstraintRecurse";
        AlterTableType[AlterTableType["AT_AddIndexConstraint"] = 26] = "AT_AddIndexConstraint";
        AlterTableType[AlterTableType["AT_DropConstraint"] = 27] = "AT_DropConstraint";
        AlterTableType[AlterTableType["AT_DropConstraintRecurse"] = 28] = "AT_DropConstraintRecurse";
        AlterTableType[AlterTableType["AT_ReAddComment"] = 29] = "AT_ReAddComment";
        AlterTableType[AlterTableType["AT_AlterColumnType"] = 30] = "AT_AlterColumnType";
        AlterTableType[AlterTableType["AT_AlterColumnGenericOptions"] = 31] = "AT_AlterColumnGenericOptions";
        AlterTableType[AlterTableType["AT_ChangeOwner"] = 32] = "AT_ChangeOwner";
        AlterTableType[AlterTableType["AT_ClusterOn"] = 33] = "AT_ClusterOn";
        AlterTableType[AlterTableType["AT_DropCluster"] = 34] = "AT_DropCluster";
        AlterTableType[AlterTableType["AT_SetLogged"] = 35] = "AT_SetLogged";
        AlterTableType[AlterTableType["AT_SetUnLogged"] = 36] = "AT_SetUnLogged";
        AlterTableType[AlterTableType["AT_DropOids"] = 37] = "AT_DropOids";
        AlterTableType[AlterTableType["AT_SetAccessMethod"] = 38] = "AT_SetAccessMethod";
        AlterTableType[AlterTableType["AT_SetTableSpace"] = 39] = "AT_SetTableSpace";
        AlterTableType[AlterTableType["AT_SetRelOptions"] = 40] = "AT_SetRelOptions";
        AlterTableType[AlterTableType["AT_ResetRelOptions"] = 41] = "AT_ResetRelOptions";
        AlterTableType[AlterTableType["AT_ReplaceRelOptions"] = 42] = "AT_ReplaceRelOptions";
        AlterTableType[AlterTableType["AT_EnableTrig"] = 43] = "AT_EnableTrig";
        AlterTableType[AlterTableType["AT_EnableAlwaysTrig"] = 44] = "AT_EnableAlwaysTrig";
        AlterTableType[AlterTableType["AT_EnableReplicaTrig"] = 45] = "AT_EnableReplicaTrig";
        AlterTableType[AlterTableType["AT_DisableTrig"] = 46] = "AT_DisableTrig";
        AlterTableType[AlterTableType["AT_EnableTrigAll"] = 47] = "AT_EnableTrigAll";
        AlterTableType[AlterTableType["AT_DisableTrigAll"] = 48] = "AT_DisableTrigAll";
        AlterTableType[AlterTableType["AT_EnableTrigUser"] = 49] = "AT_EnableTrigUser";
        AlterTableType[AlterTableType["AT_DisableTrigUser"] = 50] = "AT_DisableTrigUser";
        AlterTableType[AlterTableType["AT_EnableRule"] = 51] = "AT_EnableRule";
        AlterTableType[AlterTableType["AT_EnableAlwaysRule"] = 52] = "AT_EnableAlwaysRule";
        AlterTableType[AlterTableType["AT_EnableReplicaRule"] = 53] = "AT_EnableReplicaRule";
        AlterTableType[AlterTableType["AT_DisableRule"] = 54] = "AT_DisableRule";
        AlterTableType[AlterTableType["AT_AddInherit"] = 55] = "AT_AddInherit";
        AlterTableType[AlterTableType["AT_DropInherit"] = 56] = "AT_DropInherit";
        AlterTableType[AlterTableType["AT_AddOf"] = 57] = "AT_AddOf";
        AlterTableType[AlterTableType["AT_DropOf"] = 58] = "AT_DropOf";
        AlterTableType[AlterTableType["AT_ReplicaIdentity"] = 59] = "AT_ReplicaIdentity";
        AlterTableType[AlterTableType["AT_EnableRowSecurity"] = 60] = "AT_EnableRowSecurity";
        AlterTableType[AlterTableType["AT_DisableRowSecurity"] = 61] = "AT_DisableRowSecurity";
        AlterTableType[AlterTableType["AT_ForceRowSecurity"] = 62] = "AT_ForceRowSecurity";
        AlterTableType[AlterTableType["AT_NoForceRowSecurity"] = 63] = "AT_NoForceRowSecurity";
        AlterTableType[AlterTableType["AT_GenericOptions"] = 64] = "AT_GenericOptions";
        AlterTableType[AlterTableType["AT_AttachPartition"] = 65] = "AT_AttachPartition";
        AlterTableType[AlterTableType["AT_DetachPartition"] = 66] = "AT_DetachPartition";
        AlterTableType[AlterTableType["AT_DetachPartitionFinalize"] = 67] = "AT_DetachPartitionFinalize";
        AlterTableType[AlterTableType["AT_AddIdentity"] = 68] = "AT_AddIdentity";
        AlterTableType[AlterTableType["AT_SetIdentity"] = 69] = "AT_SetIdentity";
        AlterTableType[AlterTableType["AT_DropIdentity"] = 70] = "AT_DropIdentity";
        AlterTableType[AlterTableType["AT_ReAddStatistics"] = 71] = "AT_ReAddStatistics";
        AlterTableType[AlterTableType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(AlterTableType || (AlterTableType = {}));
    let GrantTargetType;
    (function (GrantTargetType) {
        GrantTargetType[GrantTargetType["GRANT_TARGET_TYPE_UNDEFINED"] = 0] = "GRANT_TARGET_TYPE_UNDEFINED";
        GrantTargetType[GrantTargetType["ACL_TARGET_OBJECT"] = 1] = "ACL_TARGET_OBJECT";
        GrantTargetType[GrantTargetType["ACL_TARGET_ALL_IN_SCHEMA"] = 2] = "ACL_TARGET_ALL_IN_SCHEMA";
        GrantTargetType[GrantTargetType["ACL_TARGET_DEFAULTS"] = 3] = "ACL_TARGET_DEFAULTS";
        GrantTargetType[GrantTargetType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(GrantTargetType || (GrantTargetType = {}));
    let VariableSetKind;
    (function (VariableSetKind) {
        VariableSetKind[VariableSetKind["VARIABLE_SET_KIND_UNDEFINED"] = 0] = "VARIABLE_SET_KIND_UNDEFINED";
        VariableSetKind[VariableSetKind["VAR_SET_VALUE"] = 1] = "VAR_SET_VALUE";
        VariableSetKind[VariableSetKind["VAR_SET_DEFAULT"] = 2] = "VAR_SET_DEFAULT";
        VariableSetKind[VariableSetKind["VAR_SET_CURRENT"] = 3] = "VAR_SET_CURRENT";
        VariableSetKind[VariableSetKind["VAR_SET_MULTI"] = 4] = "VAR_SET_MULTI";
        VariableSetKind[VariableSetKind["VAR_RESET"] = 5] = "VAR_RESET";
        VariableSetKind[VariableSetKind["VAR_RESET_ALL"] = 6] = "VAR_RESET_ALL";
        VariableSetKind[VariableSetKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(VariableSetKind || (VariableSetKind = {}));
    let ConstrType;
    (function (ConstrType) {
        ConstrType[ConstrType["CONSTR_TYPE_UNDEFINED"] = 0] = "CONSTR_TYPE_UNDEFINED";
        ConstrType[ConstrType["CONSTR_NULL"] = 1] = "CONSTR_NULL";
        ConstrType[ConstrType["CONSTR_NOTNULL"] = 2] = "CONSTR_NOTNULL";
        ConstrType[ConstrType["CONSTR_DEFAULT"] = 3] = "CONSTR_DEFAULT";
        ConstrType[ConstrType["CONSTR_IDENTITY"] = 4] = "CONSTR_IDENTITY";
        ConstrType[ConstrType["CONSTR_GENERATED"] = 5] = "CONSTR_GENERATED";
        ConstrType[ConstrType["CONSTR_CHECK"] = 6] = "CONSTR_CHECK";
        ConstrType[ConstrType["CONSTR_PRIMARY"] = 7] = "CONSTR_PRIMARY";
        ConstrType[ConstrType["CONSTR_UNIQUE"] = 8] = "CONSTR_UNIQUE";
        ConstrType[ConstrType["CONSTR_EXCLUSION"] = 9] = "CONSTR_EXCLUSION";
        ConstrType[ConstrType["CONSTR_FOREIGN"] = 10] = "CONSTR_FOREIGN";
        ConstrType[ConstrType["CONSTR_ATTR_DEFERRABLE"] = 11] = "CONSTR_ATTR_DEFERRABLE";
        ConstrType[ConstrType["CONSTR_ATTR_NOT_DEFERRABLE"] = 12] = "CONSTR_ATTR_NOT_DEFERRABLE";
        ConstrType[ConstrType["CONSTR_ATTR_DEFERRED"] = 13] = "CONSTR_ATTR_DEFERRED";
        ConstrType[ConstrType["CONSTR_ATTR_IMMEDIATE"] = 14] = "CONSTR_ATTR_IMMEDIATE";
        ConstrType[ConstrType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(ConstrType || (ConstrType = {}));
    let ImportForeignSchemaType;
    (function (ImportForeignSchemaType) {
        ImportForeignSchemaType[ImportForeignSchemaType["IMPORT_FOREIGN_SCHEMA_TYPE_UNDEFINED"] = 0] = "IMPORT_FOREIGN_SCHEMA_TYPE_UNDEFINED";
        ImportForeignSchemaType[ImportForeignSchemaType["FDW_IMPORT_SCHEMA_ALL"] = 1] = "FDW_IMPORT_SCHEMA_ALL";
        ImportForeignSchemaType[ImportForeignSchemaType["FDW_IMPORT_SCHEMA_LIMIT_TO"] = 2] = "FDW_IMPORT_SCHEMA_LIMIT_TO";
        ImportForeignSchemaType[ImportForeignSchemaType["FDW_IMPORT_SCHEMA_EXCEPT"] = 3] = "FDW_IMPORT_SCHEMA_EXCEPT";
        ImportForeignSchemaType[ImportForeignSchemaType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(ImportForeignSchemaType || (ImportForeignSchemaType = {}));
    let RoleStmtType;
    (function (RoleStmtType) {
        RoleStmtType[RoleStmtType["ROLE_STMT_TYPE_UNDEFINED"] = 0] = "ROLE_STMT_TYPE_UNDEFINED";
        RoleStmtType[RoleStmtType["ROLESTMT_ROLE"] = 1] = "ROLESTMT_ROLE";
        RoleStmtType[RoleStmtType["ROLESTMT_USER"] = 2] = "ROLESTMT_USER";
        RoleStmtType[RoleStmtType["ROLESTMT_GROUP"] = 3] = "ROLESTMT_GROUP";
        RoleStmtType[RoleStmtType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(RoleStmtType || (RoleStmtType = {}));
    let FetchDirection;
    (function (FetchDirection) {
        FetchDirection[FetchDirection["FETCH_DIRECTION_UNDEFINED"] = 0] = "FETCH_DIRECTION_UNDEFINED";
        FetchDirection[FetchDirection["FETCH_FORWARD"] = 1] = "FETCH_FORWARD";
        FetchDirection[FetchDirection["FETCH_BACKWARD"] = 2] = "FETCH_BACKWARD";
        FetchDirection[FetchDirection["FETCH_ABSOLUTE"] = 3] = "FETCH_ABSOLUTE";
        FetchDirection[FetchDirection["FETCH_RELATIVE"] = 4] = "FETCH_RELATIVE";
        FetchDirection[FetchDirection["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(FetchDirection || (FetchDirection = {}));
    let FunctionParameterMode;
    (function (FunctionParameterMode) {
        FunctionParameterMode[FunctionParameterMode["FUNCTION_PARAMETER_MODE_UNDEFINED"] = 0] = "FUNCTION_PARAMETER_MODE_UNDEFINED";
        FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_IN"] = 1] = "FUNC_PARAM_IN";
        FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_OUT"] = 2] = "FUNC_PARAM_OUT";
        FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_INOUT"] = 3] = "FUNC_PARAM_INOUT";
        FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_VARIADIC"] = 4] = "FUNC_PARAM_VARIADIC";
        FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_TABLE"] = 5] = "FUNC_PARAM_TABLE";
        FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_DEFAULT"] = 6] = "FUNC_PARAM_DEFAULT";
        FunctionParameterMode[FunctionParameterMode["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(FunctionParameterMode || (FunctionParameterMode = {}));
    let TransactionStmtKind;
    (function (TransactionStmtKind) {
        TransactionStmtKind[TransactionStmtKind["TRANSACTION_STMT_KIND_UNDEFINED"] = 0] = "TRANSACTION_STMT_KIND_UNDEFINED";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_BEGIN"] = 1] = "TRANS_STMT_BEGIN";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_START"] = 2] = "TRANS_STMT_START";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_COMMIT"] = 3] = "TRANS_STMT_COMMIT";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_ROLLBACK"] = 4] = "TRANS_STMT_ROLLBACK";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_SAVEPOINT"] = 5] = "TRANS_STMT_SAVEPOINT";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_RELEASE"] = 6] = "TRANS_STMT_RELEASE";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_ROLLBACK_TO"] = 7] = "TRANS_STMT_ROLLBACK_TO";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_PREPARE"] = 8] = "TRANS_STMT_PREPARE";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_COMMIT_PREPARED"] = 9] = "TRANS_STMT_COMMIT_PREPARED";
        TransactionStmtKind[TransactionStmtKind["TRANS_STMT_ROLLBACK_PREPARED"] = 10] = "TRANS_STMT_ROLLBACK_PREPARED";
        TransactionStmtKind[TransactionStmtKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(TransactionStmtKind || (TransactionStmtKind = {}));
    let ViewCheckOption;
    (function (ViewCheckOption) {
        ViewCheckOption[ViewCheckOption["VIEW_CHECK_OPTION_UNDEFINED"] = 0] = "VIEW_CHECK_OPTION_UNDEFINED";
        ViewCheckOption[ViewCheckOption["NO_CHECK_OPTION"] = 1] = "NO_CHECK_OPTION";
        ViewCheckOption[ViewCheckOption["LOCAL_CHECK_OPTION"] = 2] = "LOCAL_CHECK_OPTION";
        ViewCheckOption[ViewCheckOption["CASCADED_CHECK_OPTION"] = 3] = "CASCADED_CHECK_OPTION";
        ViewCheckOption[ViewCheckOption["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(ViewCheckOption || (ViewCheckOption = {}));
    let DiscardMode;
    (function (DiscardMode) {
        DiscardMode[DiscardMode["DISCARD_MODE_UNDEFINED"] = 0] = "DISCARD_MODE_UNDEFINED";
        DiscardMode[DiscardMode["DISCARD_ALL"] = 1] = "DISCARD_ALL";
        DiscardMode[DiscardMode["DISCARD_PLANS"] = 2] = "DISCARD_PLANS";
        DiscardMode[DiscardMode["DISCARD_SEQUENCES"] = 3] = "DISCARD_SEQUENCES";
        DiscardMode[DiscardMode["DISCARD_TEMP"] = 4] = "DISCARD_TEMP";
        DiscardMode[DiscardMode["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(DiscardMode || (DiscardMode = {}));
    let ReindexObjectType;
    (function (ReindexObjectType) {
        ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_TYPE_UNDEFINED"] = 0] = "REINDEX_OBJECT_TYPE_UNDEFINED";
        ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_INDEX"] = 1] = "REINDEX_OBJECT_INDEX";
        ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_TABLE"] = 2] = "REINDEX_OBJECT_TABLE";
        ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_SCHEMA"] = 3] = "REINDEX_OBJECT_SCHEMA";
        ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_SYSTEM"] = 4] = "REINDEX_OBJECT_SYSTEM";
        ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_DATABASE"] = 5] = "REINDEX_OBJECT_DATABASE";
        ReindexObjectType[ReindexObjectType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(ReindexObjectType || (ReindexObjectType = {}));
    let AlterTSConfigType;
    (function (AlterTSConfigType) {
        AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_TYPE_UNDEFINED"] = 0] = "ALTER_TSCONFIG_TYPE_UNDEFINED";
        AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_ADD_MAPPING"] = 1] = "ALTER_TSCONFIG_ADD_MAPPING";
        AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_ALTER_MAPPING_FOR_TOKEN"] = 2] = "ALTER_TSCONFIG_ALTER_MAPPING_FOR_TOKEN";
        AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_REPLACE_DICT"] = 3] = "ALTER_TSCONFIG_REPLACE_DICT";
        AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_REPLACE_DICT_FOR_TOKEN"] = 4] = "ALTER_TSCONFIG_REPLACE_DICT_FOR_TOKEN";
        AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_DROP_MAPPING"] = 5] = "ALTER_TSCONFIG_DROP_MAPPING";
        AlterTSConfigType[AlterTSConfigType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(AlterTSConfigType || (AlterTSConfigType = {}));
    let PublicationObjSpecType;
    (function (PublicationObjSpecType) {
        PublicationObjSpecType[PublicationObjSpecType["PUBLICATION_OBJ_SPEC_TYPE_UNDEFINED"] = 0] = "PUBLICATION_OBJ_SPEC_TYPE_UNDEFINED";
        PublicationObjSpecType[PublicationObjSpecType["PUBLICATIONOBJ_TABLE"] = 1] = "PUBLICATIONOBJ_TABLE";
        PublicationObjSpecType[PublicationObjSpecType["PUBLICATIONOBJ_TABLES_IN_SCHEMA"] = 2] = "PUBLICATIONOBJ_TABLES_IN_SCHEMA";
        PublicationObjSpecType[PublicationObjSpecType["PUBLICATIONOBJ_TABLES_IN_CUR_SCHEMA"] = 3] = "PUBLICATIONOBJ_TABLES_IN_CUR_SCHEMA";
        PublicationObjSpecType[PublicationObjSpecType["PUBLICATIONOBJ_CONTINUATION"] = 4] = "PUBLICATIONOBJ_CONTINUATION";
        PublicationObjSpecType[PublicationObjSpecType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(PublicationObjSpecType || (PublicationObjSpecType = {}));
    let AlterPublicationAction;
    (function (AlterPublicationAction) {
        AlterPublicationAction[AlterPublicationAction["ALTER_PUBLICATION_ACTION_UNDEFINED"] = 0] = "ALTER_PUBLICATION_ACTION_UNDEFINED";
        AlterPublicationAction[AlterPublicationAction["AP_AddObjects"] = 1] = "AP_AddObjects";
        AlterPublicationAction[AlterPublicationAction["AP_DropObjects"] = 2] = "AP_DropObjects";
        AlterPublicationAction[AlterPublicationAction["AP_SetObjects"] = 3] = "AP_SetObjects";
        AlterPublicationAction[AlterPublicationAction["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(AlterPublicationAction || (AlterPublicationAction = {}));
    let AlterSubscriptionType;
    (function (AlterSubscriptionType) {
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_TYPE_UNDEFINED"] = 0] = "ALTER_SUBSCRIPTION_TYPE_UNDEFINED";
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_OPTIONS"] = 1] = "ALTER_SUBSCRIPTION_OPTIONS";
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_CONNECTION"] = 2] = "ALTER_SUBSCRIPTION_CONNECTION";
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_SET_PUBLICATION"] = 3] = "ALTER_SUBSCRIPTION_SET_PUBLICATION";
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_ADD_PUBLICATION"] = 4] = "ALTER_SUBSCRIPTION_ADD_PUBLICATION";
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_DROP_PUBLICATION"] = 5] = "ALTER_SUBSCRIPTION_DROP_PUBLICATION";
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_REFRESH"] = 6] = "ALTER_SUBSCRIPTION_REFRESH";
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_ENABLED"] = 7] = "ALTER_SUBSCRIPTION_ENABLED";
        AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_SKIP"] = 8] = "ALTER_SUBSCRIPTION_SKIP";
        AlterSubscriptionType[AlterSubscriptionType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(AlterSubscriptionType || (AlterSubscriptionType = {}));
    let OnCommitAction;
    (function (OnCommitAction) {
        OnCommitAction[OnCommitAction["ON_COMMIT_ACTION_UNDEFINED"] = 0] = "ON_COMMIT_ACTION_UNDEFINED";
        OnCommitAction[OnCommitAction["ONCOMMIT_NOOP"] = 1] = "ONCOMMIT_NOOP";
        OnCommitAction[OnCommitAction["ONCOMMIT_PRESERVE_ROWS"] = 2] = "ONCOMMIT_PRESERVE_ROWS";
        OnCommitAction[OnCommitAction["ONCOMMIT_DELETE_ROWS"] = 3] = "ONCOMMIT_DELETE_ROWS";
        OnCommitAction[OnCommitAction["ONCOMMIT_DROP"] = 4] = "ONCOMMIT_DROP";
        OnCommitAction[OnCommitAction["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(OnCommitAction || (OnCommitAction = {}));
    let ParamKind;
    (function (ParamKind) {
        ParamKind[ParamKind["PARAM_KIND_UNDEFINED"] = 0] = "PARAM_KIND_UNDEFINED";
        ParamKind[ParamKind["PARAM_EXTERN"] = 1] = "PARAM_EXTERN";
        ParamKind[ParamKind["PARAM_EXEC"] = 2] = "PARAM_EXEC";
        ParamKind[ParamKind["PARAM_SUBLINK"] = 3] = "PARAM_SUBLINK";
        ParamKind[ParamKind["PARAM_MULTIEXPR"] = 4] = "PARAM_MULTIEXPR";
        ParamKind[ParamKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(ParamKind || (ParamKind = {}));
    let CoercionContext;
    (function (CoercionContext) {
        CoercionContext[CoercionContext["COERCION_CONTEXT_UNDEFINED"] = 0] = "COERCION_CONTEXT_UNDEFINED";
        CoercionContext[CoercionContext["COERCION_IMPLICIT"] = 1] = "COERCION_IMPLICIT";
        CoercionContext[CoercionContext["COERCION_ASSIGNMENT"] = 2] = "COERCION_ASSIGNMENT";
        CoercionContext[CoercionContext["COERCION_PLPGSQL"] = 3] = "COERCION_PLPGSQL";
        CoercionContext[CoercionContext["COERCION_EXPLICIT"] = 4] = "COERCION_EXPLICIT";
        CoercionContext[CoercionContext["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(CoercionContext || (CoercionContext = {}));
    let CoercionForm;
    (function (CoercionForm) {
        CoercionForm[CoercionForm["COERCION_FORM_UNDEFINED"] = 0] = "COERCION_FORM_UNDEFINED";
        CoercionForm[CoercionForm["COERCE_EXPLICIT_CALL"] = 1] = "COERCE_EXPLICIT_CALL";
        CoercionForm[CoercionForm["COERCE_EXPLICIT_CAST"] = 2] = "COERCE_EXPLICIT_CAST";
        CoercionForm[CoercionForm["COERCE_IMPLICIT_CAST"] = 3] = "COERCE_IMPLICIT_CAST";
        CoercionForm[CoercionForm["COERCE_SQL_SYNTAX"] = 4] = "COERCE_SQL_SYNTAX";
        CoercionForm[CoercionForm["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(CoercionForm || (CoercionForm = {}));
    let BoolExprType;
    (function (BoolExprType) {
        BoolExprType["BOOL_EXPR_TYPE_UNDEFINED"] = "BOOL_EXPR_TYPE_UNDEFINED";
        BoolExprType["AND_EXPR"] = "AND_EXPR";
        BoolExprType["OR_EXPR"] = "OR_EXPR";
        BoolExprType["NOT_EXPR"] = "NOT_EXPR";
        BoolExprType["UNRECOGNIZED"] = "UNRECOGNIZED";
    })(BoolExprType = LibPgQueryAST.BoolExprType || (LibPgQueryAST.BoolExprType = {}));
    let SubLinkType;
    (function (SubLinkType) {
        SubLinkType["SUB_LINK_TYPE_UNDEFINED"] = "SUB_LINK_TYPE_UNDEFINED";
        SubLinkType["EXISTS_SUBLINK"] = "EXISTS_SUBLINK";
        SubLinkType["ALL_SUBLINK"] = "ALL_SUBLINK";
        SubLinkType["ANY_SUBLINK"] = "ANY_SUBLINK";
        SubLinkType["ROWCOMPARE_SUBLINK"] = "ROWCOMPARE_SUBLINK";
        SubLinkType["EXPR_SUBLINK"] = "EXPR_SUBLINK";
        SubLinkType["MULTIEXPR_SUBLINK"] = "MULTIEXPR_SUBLINK";
        SubLinkType["ARRAY_SUBLINK"] = "ARRAY_SUBLINK";
        SubLinkType["CTE_SUBLINK"] = "CTE_SUBLINK";
        SubLinkType["UNRECOGNIZED"] = "UNRECOGNIZED";
    })(SubLinkType = LibPgQueryAST.SubLinkType || (LibPgQueryAST.SubLinkType = {}));
    let RowCompareType;
    (function (RowCompareType) {
        RowCompareType[RowCompareType["ROW_COMPARE_TYPE_UNDEFINED"] = 0] = "ROW_COMPARE_TYPE_UNDEFINED";
        RowCompareType[RowCompareType["ROWCOMPARE_LT"] = 1] = "ROWCOMPARE_LT";
        RowCompareType[RowCompareType["ROWCOMPARE_LE"] = 2] = "ROWCOMPARE_LE";
        RowCompareType[RowCompareType["ROWCOMPARE_EQ"] = 3] = "ROWCOMPARE_EQ";
        RowCompareType[RowCompareType["ROWCOMPARE_GE"] = 4] = "ROWCOMPARE_GE";
        RowCompareType[RowCompareType["ROWCOMPARE_GT"] = 5] = "ROWCOMPARE_GT";
        RowCompareType[RowCompareType["ROWCOMPARE_NE"] = 6] = "ROWCOMPARE_NE";
        RowCompareType[RowCompareType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(RowCompareType || (RowCompareType = {}));
    let MinMaxOp;
    (function (MinMaxOp) {
        MinMaxOp[MinMaxOp["MIN_MAX_OP_UNDEFINED"] = 0] = "MIN_MAX_OP_UNDEFINED";
        MinMaxOp[MinMaxOp["IS_GREATEST"] = 1] = "IS_GREATEST";
        MinMaxOp[MinMaxOp["IS_LEAST"] = 2] = "IS_LEAST";
        MinMaxOp[MinMaxOp["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(MinMaxOp || (MinMaxOp = {}));
    let SQLValueFunctionOp;
    (function (SQLValueFunctionOp) {
        SQLValueFunctionOp[SQLValueFunctionOp["SQLVALUE_FUNCTION_OP_UNDEFINED"] = 0] = "SQLVALUE_FUNCTION_OP_UNDEFINED";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_DATE"] = 1] = "SVFOP_CURRENT_DATE";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_TIME"] = 2] = "SVFOP_CURRENT_TIME";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_TIME_N"] = 3] = "SVFOP_CURRENT_TIME_N";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_TIMESTAMP"] = 4] = "SVFOP_CURRENT_TIMESTAMP";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_TIMESTAMP_N"] = 5] = "SVFOP_CURRENT_TIMESTAMP_N";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_LOCALTIME"] = 6] = "SVFOP_LOCALTIME";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_LOCALTIME_N"] = 7] = "SVFOP_LOCALTIME_N";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_LOCALTIMESTAMP"] = 8] = "SVFOP_LOCALTIMESTAMP";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_LOCALTIMESTAMP_N"] = 9] = "SVFOP_LOCALTIMESTAMP_N";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_ROLE"] = 10] = "SVFOP_CURRENT_ROLE";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_USER"] = 11] = "SVFOP_CURRENT_USER";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_USER"] = 12] = "SVFOP_USER";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_SESSION_USER"] = 13] = "SVFOP_SESSION_USER";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_CATALOG"] = 14] = "SVFOP_CURRENT_CATALOG";
        SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_SCHEMA"] = 15] = "SVFOP_CURRENT_SCHEMA";
        SQLValueFunctionOp[SQLValueFunctionOp["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(SQLValueFunctionOp || (SQLValueFunctionOp = {}));
    let XmlExprOp;
    (function (XmlExprOp) {
        XmlExprOp[XmlExprOp["XML_EXPR_OP_UNDEFINED"] = 0] = "XML_EXPR_OP_UNDEFINED";
        XmlExprOp[XmlExprOp["IS_XMLCONCAT"] = 1] = "IS_XMLCONCAT";
        XmlExprOp[XmlExprOp["IS_XMLELEMENT"] = 2] = "IS_XMLELEMENT";
        XmlExprOp[XmlExprOp["IS_XMLFOREST"] = 3] = "IS_XMLFOREST";
        XmlExprOp[XmlExprOp["IS_XMLPARSE"] = 4] = "IS_XMLPARSE";
        XmlExprOp[XmlExprOp["IS_XMLPI"] = 5] = "IS_XMLPI";
        XmlExprOp[XmlExprOp["IS_XMLROOT"] = 6] = "IS_XMLROOT";
        XmlExprOp[XmlExprOp["IS_XMLSERIALIZE"] = 7] = "IS_XMLSERIALIZE";
        XmlExprOp[XmlExprOp["IS_DOCUMENT"] = 8] = "IS_DOCUMENT";
        XmlExprOp[XmlExprOp["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(XmlExprOp || (XmlExprOp = {}));
    let XmlOptionType;
    (function (XmlOptionType) {
        XmlOptionType[XmlOptionType["XML_OPTION_TYPE_UNDEFINED"] = 0] = "XML_OPTION_TYPE_UNDEFINED";
        XmlOptionType[XmlOptionType["XMLOPTION_DOCUMENT"] = 1] = "XMLOPTION_DOCUMENT";
        XmlOptionType[XmlOptionType["XMLOPTION_CONTENT"] = 2] = "XMLOPTION_CONTENT";
        XmlOptionType[XmlOptionType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(XmlOptionType || (XmlOptionType = {}));
    let NullTestType;
    (function (NullTestType) {
        NullTestType["NULL_TEST_TYPE_UNDEFINED"] = "NULL_TEST_TYPE_UNDEFINED";
        NullTestType["IS_NULL"] = "IS_NULL";
        NullTestType["IS_NOT_NULL"] = "IS_NOT_NULL";
        NullTestType["UNRECOGNIZED"] = "UNRECOGNIZED";
    })(NullTestType = LibPgQueryAST.NullTestType || (LibPgQueryAST.NullTestType = {}));
    let BoolTestType;
    (function (BoolTestType) {
        BoolTestType[BoolTestType["BOOL_TEST_TYPE_UNDEFINED"] = 0] = "BOOL_TEST_TYPE_UNDEFINED";
        BoolTestType[BoolTestType["IS_TRUE"] = 1] = "IS_TRUE";
        BoolTestType[BoolTestType["IS_NOT_TRUE"] = 2] = "IS_NOT_TRUE";
        BoolTestType[BoolTestType["IS_FALSE"] = 3] = "IS_FALSE";
        BoolTestType[BoolTestType["IS_NOT_FALSE"] = 4] = "IS_NOT_FALSE";
        BoolTestType[BoolTestType["IS_UNKNOWN"] = 5] = "IS_UNKNOWN";
        BoolTestType[BoolTestType["IS_NOT_UNKNOWN"] = 6] = "IS_NOT_UNKNOWN";
        BoolTestType[BoolTestType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(BoolTestType || (BoolTestType = {}));
    let CmdType;
    (function (CmdType) {
        CmdType[CmdType["CMD_TYPE_UNDEFINED"] = 0] = "CMD_TYPE_UNDEFINED";
        CmdType[CmdType["CMD_UNKNOWN"] = 1] = "CMD_UNKNOWN";
        CmdType[CmdType["CMD_SELECT"] = 2] = "CMD_SELECT";
        CmdType[CmdType["CMD_UPDATE"] = 3] = "CMD_UPDATE";
        CmdType[CmdType["CMD_INSERT"] = 4] = "CMD_INSERT";
        CmdType[CmdType["CMD_DELETE"] = 5] = "CMD_DELETE";
        CmdType[CmdType["CMD_MERGE"] = 6] = "CMD_MERGE";
        CmdType[CmdType["CMD_UTILITY"] = 7] = "CMD_UTILITY";
        CmdType[CmdType["CMD_NOTHING"] = 8] = "CMD_NOTHING";
        CmdType[CmdType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(CmdType || (CmdType = {}));
    let JoinType;
    (function (JoinType) {
        JoinType["JOIN_TYPE_UNDEFINED"] = "JOIN_TYPE_UNDEFINED";
        JoinType["JOIN_INNER"] = "JOIN_INNER";
        JoinType["JOIN_LEFT"] = "JOIN_LEFT";
        JoinType["JOIN_FULL"] = "JOIN_FULL";
        JoinType["JOIN_RIGHT"] = "JOIN_RIGHT";
        JoinType["JOIN_SEMI"] = "JOIN_SEMI";
        JoinType["JOIN_ANTI"] = "JOIN_ANTI";
        JoinType["JOIN_UNIQUE_OUTER"] = "JOIN_UNIQUE_OUTER";
        JoinType["JOIN_UNIQUE_INNER"] = "JOIN_UNIQUE_INNER";
        JoinType["UNRECOGNIZED"] = "UNRECOGNIZED";
    })(JoinType = LibPgQueryAST.JoinType || (LibPgQueryAST.JoinType = {}));
    let AggSplit;
    (function (AggSplit) {
        AggSplit[AggSplit["AGG_SPLIT_UNDEFINED"] = 0] = "AGG_SPLIT_UNDEFINED";
        AggSplit[AggSplit["AGGSPLIT_SIMPLE"] = 1] = "AGGSPLIT_SIMPLE";
        AggSplit[AggSplit["AGGSPLIT_INITIAL_SERIAL"] = 2] = "AGGSPLIT_INITIAL_SERIAL";
        AggSplit[AggSplit["AGGSPLIT_FINAL_DESERIAL"] = 3] = "AGGSPLIT_FINAL_DESERIAL";
        AggSplit[AggSplit["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(AggSplit || (AggSplit = {}));
    let OnConflictAction;
    (function (OnConflictAction) {
        OnConflictAction[OnConflictAction["ON_CONFLICT_ACTION_UNDEFINED"] = 0] = "ON_CONFLICT_ACTION_UNDEFINED";
        OnConflictAction[OnConflictAction["ONCONFLICT_NONE"] = 1] = "ONCONFLICT_NONE";
        OnConflictAction[OnConflictAction["ONCONFLICT_NOTHING"] = 2] = "ONCONFLICT_NOTHING";
        OnConflictAction[OnConflictAction["ONCONFLICT_UPDATE"] = 3] = "ONCONFLICT_UPDATE";
        OnConflictAction[OnConflictAction["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(OnConflictAction || (OnConflictAction = {}));
    let LimitOption;
    (function (LimitOption) {
        LimitOption[LimitOption["LIMIT_OPTION_UNDEFINED"] = 0] = "LIMIT_OPTION_UNDEFINED";
        LimitOption[LimitOption["LIMIT_OPTION_DEFAULT"] = 1] = "LIMIT_OPTION_DEFAULT";
        LimitOption[LimitOption["LIMIT_OPTION_COUNT"] = 2] = "LIMIT_OPTION_COUNT";
        LimitOption[LimitOption["LIMIT_OPTION_WITH_TIES"] = 3] = "LIMIT_OPTION_WITH_TIES";
        LimitOption[LimitOption["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(LimitOption || (LimitOption = {}));
    let LockClauseStrength;
    (function (LockClauseStrength) {
        LockClauseStrength[LockClauseStrength["LOCK_CLAUSE_STRENGTH_UNDEFINED"] = 0] = "LOCK_CLAUSE_STRENGTH_UNDEFINED";
        LockClauseStrength[LockClauseStrength["LCS_NONE"] = 1] = "LCS_NONE";
        LockClauseStrength[LockClauseStrength["LCS_FORKEYSHARE"] = 2] = "LCS_FORKEYSHARE";
        LockClauseStrength[LockClauseStrength["LCS_FORSHARE"] = 3] = "LCS_FORSHARE";
        LockClauseStrength[LockClauseStrength["LCS_FORNOKEYUPDATE"] = 4] = "LCS_FORNOKEYUPDATE";
        LockClauseStrength[LockClauseStrength["LCS_FORUPDATE"] = 5] = "LCS_FORUPDATE";
        LockClauseStrength[LockClauseStrength["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(LockClauseStrength || (LockClauseStrength = {}));
    let LockWaitPolicy;
    (function (LockWaitPolicy) {
        LockWaitPolicy[LockWaitPolicy["LOCK_WAIT_POLICY_UNDEFINED"] = 0] = "LOCK_WAIT_POLICY_UNDEFINED";
        LockWaitPolicy[LockWaitPolicy["LockWaitBlock"] = 1] = "LockWaitBlock";
        LockWaitPolicy[LockWaitPolicy["LockWaitSkip"] = 2] = "LockWaitSkip";
        LockWaitPolicy[LockWaitPolicy["LockWaitError"] = 3] = "LockWaitError";
        LockWaitPolicy[LockWaitPolicy["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(LockWaitPolicy || (LockWaitPolicy = {}));
    let KeywordKind;
    (function (KeywordKind) {
        KeywordKind[KeywordKind["NO_KEYWORD"] = 0] = "NO_KEYWORD";
        KeywordKind[KeywordKind["UNRESERVED_KEYWORD"] = 1] = "UNRESERVED_KEYWORD";
        KeywordKind[KeywordKind["COL_NAME_KEYWORD"] = 2] = "COL_NAME_KEYWORD";
        KeywordKind[KeywordKind["TYPE_FUNC_NAME_KEYWORD"] = 3] = "TYPE_FUNC_NAME_KEYWORD";
        KeywordKind[KeywordKind["RESERVED_KEYWORD"] = 4] = "RESERVED_KEYWORD";
        KeywordKind[KeywordKind["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
    })(KeywordKind || (KeywordKind = {}));
})(LibPgQueryAST = exports.LibPgQueryAST || (exports.LibPgQueryAST = {}));
